qiskit/transpiler/passes/optimization/consolidate_blocks.py

```diff
@@ -159,11 +167,37 @@ def run(self, dag):
                         dag.remove_op_node(node)
                 else:
                     dag.replace_block_with_op(run, unitary, {qubit: 0}, cycle_check=False)
+
+        dag = self._handle_control_flow_ops(dag)
+
         # Clear collected blocks and runs as they are no longer valid after consolidation
         if "run_list" in self.property_set:
             del self.property_set["run_list"]
         if "block_list" in self.property_set:
             del self.property_set["block_list"]
+
+        return dag
+
+    def _handle_control_flow_ops(self, dag):
+        """
+        This is similar to transpiler/passes/utils/control_flow.py except that the
+        collect blocks is redone for the control flow blocks.
+        """
+        from qiskit.transpiler import PassManager
+
+        pass_manager = PassManager()
+        if "run_list" in self.property_set:
+            pass_manager.append(Collect1qRuns())
+        if "block_list" in self.property_set:
+            pass_manager.append(Collect2qBlocks())
```

Comment:
We can likely make this `PassManager` just once up in `__init__` and retrieve it from `self` each time, even during the recursion.

Reply:
Good idea. The cost should be negligible.

Reply:
After trying, I don't see an easy way to make this work. A `ConsolidateBlocks`  instance is constructed before the top-level  `RunningPassManager` injects the necessary information. The call `PassManager` on my machine takes < 250ns. Constructing it and checking a `dict` and appending passes takes 13us or 26us.

We could pass the required information when constructing `ConsolidateBlocks`. That would allow us to do this more cleanly in the future. Maybe populate `requires` based boolean-valued arguments. We'd support the status quo for a while.

In fact, doing this would be less fragile and better signal intent than the status quo. This:

https://github.com/Qiskit/qiskit-terra/blob/fb9d5d8bb41f1e289b5ee895ea087cc92e74a921/qiskit/transpiler/preset_passmanagers/level3.py#L174-L179

is not as transparent as
```python
        ConsolidateBlocks(
            collect_2q_blocks=True, basis_gates=basis_gates, target=target, approximation_degree=approximation_degree
        ),
        UnitarySynthesis(
            basis_gates,
```

Reply:
Passing the required information to `ConsolidateBlocks` would be my general choice so that the top level of the recursion and the recursive calls can all look more similar. At this stage in the 0.25 cycle, it might be better to leave what you've got and fix it in 0.45 instead so we've got more time.